src/data/energy.ts

```diff
@@ -128,3 +133,138 @@ export const energySourcesByType = (prefs: EnergyPreferences) => {
   }
   return types;
 };
+
+export interface EnergyData {
+  start: Date;
+  end?: Date;
+  prefs: EnergyPreferences;
+  info: EnergyInfo;
+  stats: Statistics;
+  co2SignalConfigEntry?: ConfigEntry;
+  co2SignalEntity?: string;
+}
+
+export const getEnergyData = async (
+  hass: HomeAssistant,
+  prefs: EnergyPreferences,
+  start: Date,
+  end?: Date
+): Promise<EnergyData> => {
+  const [configEntries, entityRegistryEntries, info] = await Promise.all([
+    getConfigEntries(hass),
+    subscribeOne(hass.connection, subscribeEntityRegistry),
+    getEnergyInfo(hass),
+  ]);
+
+  const co2SignalConfigEntry = configEntries.find(
+    (entry) => entry.domain === "co2signal"
+  );
+
+  let co2SignalEntity: string | undefined;
+
+  if (co2SignalConfigEntry) {
+    for (const entry of entityRegistryEntries) {
+      if (entry.config_entry_id !== co2SignalConfigEntry.entry_id) {
+        continue;
+      }
+
+      // The integration offers 2 entities. We want the % one.
+      const co2State = hass.states[entry.entity_id];
+      if (!co2State || co2State.attributes.unit_of_measurement !== "%") {
+        continue;
+      }
+
+      co2SignalEntity = co2State.entity_id;
+      break;
+    }
+  }
+
+  const statIDs: string[] = [];
+
+  if (co2SignalEntity !== undefined) {
+    statIDs.push(co2SignalEntity);
+  }
+
+  for (const source of prefs.energy_sources) {
+    if (source.type === "solar") {
+      statIDs.push(source.stat_energy_from);
+      continue;
+    }
+
+    // grid source
+    for (const flowFrom of source.flow_from) {
+      statIDs.push(flowFrom.stat_energy_from);
+    }
+    for (const flowTo of source.flow_to) {
+      statIDs.push(flowTo.stat_energy_to);
+    }
+  }
+
+  const stats = await fetchStatistics(hass!, start, end, statIDs);
+
+  return {
+    start,
+    end,
+    info,
+    prefs,
+    stats,
+    co2SignalConfigEntry,
+    co2SignalEntity,
+  };
+};
+
+export interface EnergyCollection extends Collection<EnergyData> {
+  prefs?: EnergyPreferences;
+  clearPrefs(): void;
```

Comment:
Call this if you want to make sure that the next refresh uses the new preferences.